crates/ruff/src/rules/ruff/rules/invalid_index_type.rs

```diff
@@ -0,0 +1,176 @@
+                    if let Expr::Constant(ExprConstant {
+                        value: index_value, ..
+                    }) = slice_bound.as_ref()
+                    {
+                        if !(index_value.is_int() || index_value.is_none()) {
+                            checker.diagnostics.push(Diagnostic::new(
+                                InvalidIndexType {
+                                    var_type: expression_type_name(value)
+                                        .expect("Failed to cast parent expression to type name"),
+                                    idx_type: constant_type_name(index_value),
+                                    slice_bound: true,
+                                },
+                                slice_bound.range(),
+                            ));
+                        }
+                    } else if matches!(
+                        slice_bound.as_ref(),
+                        Expr::Tuple(_)
+                            | Expr::List(_)
+                            | Expr::Set(_)
+                            | Expr::Dict(_)
+                            | Expr::ListComp(_)
+                            | Expr::DictComp(_)
+                            | Expr::JoinedStr(_)
+                    ) {
+                        checker.diagnostics.push(Diagnostic::new(
+                            InvalidIndexType {
+                                var_type: expression_type_name(value)
+                                    .expect("Failed to cast parent expression to type name"),
+                                idx_type: expression_type_name(slice_bound)
+                                    .expect("Failed to slice bound expression to type name"),
+                                slice_bound: true,
+                            },
+                            slice_bound.range(),
+                        ));
+                    }
+                }
+            }
+            // If it's some other data type, it's a violation
+            Expr::Tuple(_)
+            | Expr::List(_)
+            | Expr::Set(_)
+            | Expr::Dict(_)
+            | Expr::ListComp(_)
+            | Expr::DictComp(_)
+            | Expr::JoinedStr(_) => {
+                checker.diagnostics.push(Diagnostic::new(
+                    InvalidIndexType {
+                        var_type: expression_type_name(value)
+                            .expect("Failed to cast parent expression to type name"),
+                        idx_type: expression_type_name(slice)
+                            .expect("Failed to cast index expression to type name"),
+                        slice_bound: false,
+                    },
+                    slice.range(),
+                ));
+            }
+            // If it's anything else, it's too hard to tell if it's a violation
+            _ => (),
+        }
+    }
+}
+
+fn constant_type_name(constant: &Constant) -> String {
```

Comment:
Can this return `&'static str`? Same with `expression_type_name`. The allocation (`.to_string()`) shouldn't be necessary since it's just returning a constant value.

Reply:
Definitely, but won't I need to allocate `to_string()` when I place it in the `InvalidIndexType` struct?

Reply:
Yes! But it's best to defer that allocation until it's necessary (both in practice and just as a matter of API design), both because it lets you avoid allocations in some cases (maybe the caller doesn't need a `String`), and because you can always convert a `&str` to `String` but not other way around (so returning `&str` is more flexible). In this case, you'd call `.to_string()` on the returned value of this method.

Reply:
👍 thanks!